ubuntu_image/image.py

```diff
@@ -69,52 +85,89 @@ def copy_blob(self, blob_path, **dd_args):
 
-    def partition(self, partnum, **sgdisk_args):
-        """Manipulate the GPT contained in the image file.
+    def partition(self, offset, size, name=None, is_bootable=False):
+        """Add a new partition in the image file.
+
+        The newly added partition will be appended to the existing partition
+        table on the image as defined by the volume schema.  This is all done
+        by pyparted.  Please note that libparted has no means of changing the
+        partition type GUID directly (this can only be done by setting
+        partition flags) so this has to be done separately after *all*
+        partitions have been added.
+        Also, the commit() operation also clobbers the hybrid MBR in GPT labels
+        so be sure to first perform partitioning and only afterwards attempting
+        copy operations.
+
+        :param offset: Offset (start position) of the partition in bytes.
+        :type offset: int
+        :param size: Size of partition in bytes.
+        :type size: int
+        :param name: Name of the partition.
+        :type name: str
+        :param is_bootable: Toggle if the bootable flag should be set.
+        :type name: bool
 
-        The manipulation is done using ``sgdisk`` for consistency.  The
-        device operated on is the image file represented by this
-        instance.  The keyword arguments are passed directly to the
-        ``sgdisk`` call (after tweaking to prefix the keys with ``--``
-        for the command line switch syntax).  See the sgdisk(8) manpage
-        for details.
+        """
+        # When defining geometries for our partitions we can't use the pyparted
+        # parted.sizeToSectors() function as it actually rounds down the sizes
+        # instead of rounding up, which means you might end up not having
+        # enough sectors for a partition's contents (LP: #1661298).
+        geometry = parted.Geometry(
+            device=self.device,
+            start=ceil(offset / self.sector_size),
+            length=ceil(size / self.sector_size))
+        partition = parted.Partition(
+            disk=self.disk,
+            type=parted.PARTITION_NORMAL,
+            geometry=geometry)
+        # Force an exact geometry constraint as otherwise libparted tries to be
+        # too smart and changes our geometry itself.
+        constraint = parted.Constraint(
+            exactGeom=geometry)
+        self.disk.addPartition(partition, constraint)
+        # Sadly the current pyparted bindings do not export a setter for the
+        # name of a partition (LP: #1661297).  To work-around this we need to
+        # reach out to the internal PedPartition object of the partition to
+        # call the set_name() function.
+        # We also follow the same guideline as before - for mbr labels we just
+        # ignore the name as it's not supported.
+        if name and self.schema is not VolumeSchema.mbr:
+            partition._Partition__partition.set_name(name)
```

Comment:
Youch!  But as you say it's the only way currently.  Can you put an XXX comment right before this line so we can fix it at some point in the future?